craft_parts/executor/part_handler.py

```diff
@@ -73,34 +78,148 @@ def __call__(
     ) -> None: ...
 
 
+class LayoutItem:
+    """LayoutItem maps a mountpoint to a craft partition."""
+
+    mount: str
+    partition: str
+
+    def __init__(self, mount: str, partition: str) -> None:
+        self.mount = mount
+        self.partition = partition
+
+
+DEFAULT_PARTITION = "default"
+
+# Define a harcoded layout to prototype
+# Items in the list must be ordered by mount depth, deeper first.
+prototype_layout: list[LayoutItem] = [
+    LayoutItem(partition="volume/pc/efi", mount="/boot/efi/"),
+    LayoutItem(partition=DEFAULT_PARTITION, mount="/"),
+]
+
+
+# map the source path to the destination path in the partition
+_Migrated_Contents = dict[str, str]
+
+
 class _Squasher:
     """A helper to squash layers and migrate layered content."""
 
-    def __init__(self) -> None:
-        self.migrated_files: set[str] = set()
-        self.migrated_dirs: set[str] = set()
+    def __init__(self, partition: str, layout: list[LayoutItem] | None = None) -> None:
+        self.migrated_files: dict[str, _Migrated_Contents] = {partition: {}}
+        self.migrated_directories: dict[str, _Migrated_Contents] = {partition: {}}
+        self._src_partition = partition
+        self._layout: list[LayoutItem] = [LayoutItem(partition=partition, mount="/")]
+        if layout:
+            self._layout: list[LayoutItem] = layout
 
     def migrate(
+        self,
+        refdir: Path,
+        srcdir: Path,
+        destdirs: Mapping[str | None, Path],
+        permissions: list[Permissions] | None = None,
+    ) -> None:
+        """Migrate layered content from a partition to destination directories.
+
+        If the source partition is the default one, content can be distributed to other
+        partitions using the provided layout.
+        """
+        if self._src_partition == DEFAULT_PARTITION:
+            # Distribute content into partitions according to the layout
+            for entry in self._layout:
+                # Only migrate content from the subdirectory indicated by the layout entry
+                sub_path = entry.mount.lstrip("/")
```

Comment:
`sub_path` is the mount point? Maybe call it `mountpoint_path`?

Reply:
It is indeed the mount point but I suspect that using this name could lead to confusion (especially given the conversation we had last week, showing that this mount point could mean different things based on the context). 
That could be fine to use this name here but I am not sure `_migrate()` should be aware of this, so I opted for a more generic `sub_path` name.